flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java

```diff
@@ -0,0 +1,780 @@
+
+		if (getStateBackendSerializationTimeliness() == BackendSerializationTimeliness.ON_ACCESS) {
+			assertEquals((Integer) 2, CustomStringSerializer.serializeCalled.get(SerializerVersion.NEW));
+			assertEquals((Integer) 3, CustomStringSerializer.deserializeCalled.get(SerializerVersion.NEW));
+		} else {
+			assertEquals(null, CustomStringSerializer.serializeCalled.get(SerializerVersion.NEW));
+			assertEquals(null, CustomStringSerializer.deserializeCalled.get(SerializerVersion.NEW));
+		}
+		CustomStringSerializer.resetCountingMaps();
+
+		// ============ Snapshot #2 ============
+		//  For eager serialization backends:
+		//    This should not result in any serialize / deserialize calls
+		//
+		//  For lazy serialization backends:
+		//    This should result in serializer personality NEW having 2 serialize calls
+		KeyedStateHandle snapshot2 = runSnapshot(
+			backend.snapshot(2L, 3L, streamFactory, CheckpointOptions.forCheckpointWithDefaultLocation()),
+			sharedStateRegistry);
+		backend.dispose();
+
+		if (getStateBackendSerializationTimeliness() == BackendSerializationTimeliness.ON_ACCESS) {
+			assertEquals(null, CustomStringSerializer.serializeCalled.get(SerializerVersion.NEW));
+			assertEquals(null, CustomStringSerializer.deserializeCalled.get(SerializerVersion.NEW));
+		} else {
+			assertEquals((Integer) 2, CustomStringSerializer.serializeCalled.get(SerializerVersion.NEW));
+			assertEquals(null, CustomStringSerializer.deserializeCalled.get(SerializerVersion.NEW));
+		}
+		CustomStringSerializer.resetCountingMaps();
+
+		// and restore once with NEW from NEW so that we see a read using the NEW serializer
+		// on the file backend
+		ValueStateDescriptor<String> newKvId2 = new ValueStateDescriptor<>(
+			"id",
+			new CustomStringSerializer(SerializerCompatibilityType.COMPATIBLE_AS_IS, SerializerVersion.NEW));
+
+		// ============ Restore from snapshot #2 ============
+		//  For eager serialization backends:
+		//    This should not result in any serialize / deserialize calls
+		//
+		//  For lazy serialization backends:
+		//    This should result in serializer personality RESTORE having 2 deserialize calls
+		backend = restoreKeyedBackend(IntSerializer.INSTANCE, snapshot2);
+
+		if (getStateBackendSerializationTimeliness() == BackendSerializationTimeliness.ON_ACCESS) {
+			assertEquals(null, CustomStringSerializer.serializeCalled.get(SerializerVersion.RESTORE));
+			assertEquals(null, CustomStringSerializer.deserializeCalled.get(SerializerVersion.RESTORE));
+		} else {
+			assertEquals(null, CustomStringSerializer.serializeCalled.get(SerializerVersion.RESTORE));
+			assertEquals((Integer) 2, CustomStringSerializer.deserializeCalled.get(SerializerVersion.RESTORE));
+		}
+		CustomStringSerializer.resetCountingMaps();
+
+		backend.getPartitionedState(VoidNamespace.INSTANCE, CustomVoidNamespaceSerializer.INSTANCE, newKvId2);
+		snapshot2.discardState();
+		snapshot1.discardState();
+
+		backend.dispose();
```

Comment:
In general, whenever an exception can happen before disposing a backend I would suggest to do it in a `finally` clause because a test that fails with exception can leave unclosed native resources that corrupt/segfault the JVM if it is reused for more tests after the failure.